Add representative h-card parsing, fix documentation autogeneration

capjamesg opened the pull request:
This PR introdcues a new representative h-card parsing function. This function implements the algorithm of the same name, used to find the "main" h-card on a page.

This PR also includes a number of structural changes so we can auto-generate a lot of our documentation. These changes are limited only to adding new imports to files.

Our documentation currently operates under the assumption that all modules should be generated with this syntax:

indieweb_utils.function_name()

No folder names (i.e. "indieauth") are yet needed. If we want to add those folder names, we can reject this PR and restructure our documentation as appropriate.

For illustration, this PR includes one example of an inline code snippet in a docstring for the get_post_type() function. If we approve of how this shows in the documentation, we could start adding similar snippets to the rest of the codebase to make it easier for developers to use the library.

capjamesg committed:
add representative h card parsing

capjamesg committed:
fix linting error, document new function

capjamesg committed:
finish auto generation for indieauth docs, add new imports

capjamesg committed:
add example code snippet

capjamesg committed:
add example code snippet

capjamesg committed:
add example code snippet

capjamesg committed:
add new imports to allow for documentation to be generated correclty

capjamesg committed:
minor changes to docs

capjamesg committed:
add representative h card test case

capjamesg committed:
fix merge conflict

jamesvandyne commented on src/indieweb_utils/posts/representative_h_card.py:
```diff
@@ -0,0 +1,41 @@
+    pass
+
+
+def get_representative_h_card(url: str) -> dict:
```
```suggestion
def get_representative_h_card(url: str) -> Dict[str, Any]
```

jamesvandyne commented on tests/test_posts/test_representative_h_card.py:
```diff
@@ -0,0 +1,14 @@
+        return get_representative_h_card
+
+    def test_representative_h_card(self, target):
+        url = "https://aaronparecki.com"
```
This test is making a network call and should `responses` like other tests to prevent network access when running tests.

jamesvandyne reviewed (approved):
LGTM!

jamesvandyne commented:
> No folder names (i.e. "indieauth") are yet needed. If we want to add those folder names, we can reject this PR and restructure our documentation as appropriate.

I think that's fine for now. We may want to add in specific groupings for documentation later (e.g. flask or Django example / snippets), but we can handle those documentation changes at that time.

capjamesg committed:
address linting issues, add new type hint

capjamesg commented:
I have addressed the linting issues with this PR and made the type hint change mentioned earlier.

@jamesvandyne Could you assist with putting together the mock request? I have not yet done this so would love an example to which I can refer that uses an article body.

jamesvandyne commented:
> @jamesvandyne Could you assist with putting together the mock request? I have not yet done this so would love an example to which I can refer that uses an article body.

@capjamesg There's some examples in the existing tests [test_indieauth/test_profile.py](https://github.com/capjamesg/indieweb-utils/blob/main/tests/test_indieauth/test_profile.py). But the basic idea is that we pass a crafted html string to confirm we're parsing as expected. It should look something like this...

```python
import resposnes

@responses.activate  # activates respones
def test_gets_h_card():
    url = "http://example.com"
    # build a Response object and add it to the responses inventory for this test.
    # responses will return the defined body (or status_code/headers) for any request that matches
    # the method (uppercase) and url
    responses.add(
        responses.Response( 
            method="GET",
            url=url,
            body="""<html><body><!-- add the html we want to test here and make sure we can extract properly -->"""
        )
    )
   actual = target(url)
   assert actual["some_expected_key"] == "expected_value"
```

capjamesg committed:
add representative h-card test

capjamesg committed:
Merge branch 'main' into new_docs

capjamesg commented:
I have fixed all of the merge conflicts associated with the changes in this branch. I have also added the requisite h-card test using responses rather than a HTTP request. I am going to merge this change into master upon completion of all CI checks. After merging, we should finally have documentation with all of the code snippets showing up in the docs as expected.

capjamesg committed:
merge with main

capjamesg committed:
add changes to ensure relmeauth tests pass on this branch
